rmf_site_editor/src/site/scenario.rs

```diff
@@ -102,65 +135,214 @@ pub fn update_current_scenario(
+/// Checks that the current scenario's included instances are categorized correctly
+pub fn handle_instance_updates(
+    mut commands: Commands,
+    mut change_current_scenario: EventWriter<ChangeCurrentScenario>,
+    mut scenario_instances: Query<(&mut Instance, &Affiliation<Entity>)>,
+    mut update_instance: EventReader<UpdateInstance>,
+    children: Query<&Children>,
+    current_scenario: Res<CurrentScenario>,
+    scenarios: Query<(Entity, &mut Scenario<Entity>)>,
+    model_instances: Query<&NameInSite, With<InstanceMarker>>,
+    recall_instance: Query<&RecallInstance>,
+) {
+    for update in update_instance.read() {
+        let parent_pose = retrieve_parent_pose(
+            update.instance,
+            update.scenario,
+            &children,
+            &recall_instance,
+            &scenarios,
+            &scenario_instances,
+        );
+        match update.update_type {
+            UpdateInstanceType::Add(new_pose) => {
+                commands
+                    .spawn(Instance::new_added(new_pose.clone()))
+                    .insert(Affiliation(Some(update.instance)))
+                    .set_parent(update.scenario);
+                // Insert this new instance into children scenarios as Inherited
+                let mut subtree_dependents = HashSet::<Entity>::new();
+                let mut queue = vec![update.scenario];
+                while let Some(scenario_entity) = queue.pop() {
+                    if let Ok(children) = children.get(scenario_entity) {
+                        children.iter().for_each(|e| {
+                            subtree_dependents.insert(*e);
+                            queue.push(*e);
+                        });
+                    }
+                }
+                // Only insert new instance in children scenarios. Other parent/root
+                // scenarios will not have access to this instance
+                for dependent in subtree_dependents.drain() {
+                    if let Ok((child_entity, _)) = scenarios.get(dependent) {
+                        commands
+                            .spawn(Instance::new_inherited(None))
+                            .insert(Affiliation(Some(update.instance)))
+                            .set_parent(child_entity);
+                    }
                 }
+            }
+            _ => {
+                let instance_entities =
+                    get_scenario_instance_entities(update.scenario, &children, &scenario_instances);
+                let Some(((mut instance, _), scenario_child)) = instance_entities
+                    .iter()
+                    .find(|(_, i)| *i == update.instance)
+                    .and_then(|(c_entity, _)| {
+                        scenario_instances
+                            .get_mut(*c_entity)
+                            .ok()
+                            .zip(Some(c_entity))
+                    })
+                else {
+                    continue;
```

Comment:
It looks like this will short circuit if the instance modifier doesn't already exist for this scenario. That doesn't seem robust to me. I would think we should insert it if it doesn't already exist. I realize that's tricky since insertion commands don't mesh well with `&mut T`.